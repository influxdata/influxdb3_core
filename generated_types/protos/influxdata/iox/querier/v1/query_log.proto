syntax = "proto3";
package influxdata.iox.querier.v1;
option go_package = "github.com/influxdata/iox/querier/v1";

import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";

import "influxdata/iox/querier/v1/params.proto";

service QueryLogService {
    // Get the process-local query log.
    rpc GetLog(GetLogRequest) returns (stream GetLogResponse);
}

message GetLogRequest {}

// Describe a single query.
message LogEntry {
  // Unique entry ID.
  string id = 1;

  // Current lifecycle phase of the query.
  Phase phase = 16;

  // Namespace ID.
  int64 namespace_id = 2;

  // Namespace name.
  string namespace_name = 3;

  // Query type, e.g. `sql`.
  string query_type = 4;

  // Query text.
  string query_text = 5;

  // Query parameters
  repeated QueryParam query_params = 21;

  // Trace ID.
  string trace_id = 6;

  // Start timestamp.
  google.protobuf.Timestamp issue_time = 7;

  // Number of partitions the query was executed on.
  int64 partitions = 18;

  // Number of parquet files the query was executed on.
  int64 parquet_files = 19;

  reserved 8;
  reserved 'query_completed_duration';

  // Duration it took to acquire a semaphore permit, relative to `issue_time`.
  google.protobuf.Duration permit_duration = 10;

  // Duration it took to plan the query, relative to `issue_time` + `permit_duration`.
  google.protobuf.Duration plan_duration = 11;

  // Duration it took to execute the query, relative to `issue_time` +
  // `permit_duration` + `plan_duration`.
  google.protobuf.Duration execute_duration = 12;

  // Duration from `issue_time` til the query ended somehow.
  google.protobuf.Duration end2end_duration = 13;

  // CPU duration spend for computation.
  google.protobuf.Duration compute_duration = 15;

  // Maximum memory allocated when executing the query.
  int64 max_memory = 20;

  // If the query completed successfully.
  bool success = 9;

  // If the query is currently running (in any state).
  bool running = 14;

  // If the query was cancelled.
  bool cancelled = 17;
}

// Metadata for the entire response.
message Metadata {
  // Maximum size of the query log.
  uint64 max_size = 1;

  // Number of evicted entries due to the "max size" constraint.
  uint64 evicted = 2;
}

message GetLogResponse {
  oneof data {
    Metadata metadata = 1;
    LogEntry entry = 2;
  }
}

enum Phase {
  PHASE_UNSPECIFIED = 0;
  PHASE_RECEIVED = 1;
  PHASE_PLANNED = 2;
  PHASE_PERMIT = 3;
  PHASE_CANCEL = 4;
  PHASE_SUCCESS = 5;
  PHASE_FAIL = 6;
}
