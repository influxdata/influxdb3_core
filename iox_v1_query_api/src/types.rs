use arrow::datatypes::SchemaRef;
use datafusion::physical_plan::SendableRecordBatchStream;
use serde::{Deserialize, Serialize};

use iox_query::query_log::PermitAndToken;

/// UNIX epoch precision.
/// Doc: <https://docs.influxdata.com/influxdb/v1/query_language/spec/#duration-units>
#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq)]
pub enum Precision {
    #[serde(rename = "ns")]
    Nanoseconds,
    #[serde(rename = "u", alias = "µ")]
    Microseconds,
    #[serde(rename = "ms")]
    Milliseconds,
    #[serde(rename = "s")]
    Seconds,
    #[serde(rename = "m")]
    Minutes,
    #[serde(rename = "h")]
    Hours,
    #[serde(rename = "d")]
    Days,
    #[serde(rename = "w")]
    Weeks,
}

/// An executing InfluxQL statement that produces results that can be
/// streamed as CSV.
pub(crate) struct Statement {
    pub schema: SchemaRef,
    pub permit_state: PermitAndToken,
    pub stream: SendableRecordBatchStream,
}

impl Statement {
    pub(crate) fn new(
        schema: SchemaRef,
        permit_state: PermitAndToken,
        stream: SendableRecordBatchStream,
    ) -> Self {
        Self {
            schema,
            permit_state,
            stream,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use serde_json::{from_str, to_string};

    #[test]
    fn test_precision_deserialize() {
        // Test standard rename attributes
        assert_eq!(
            from_str::<Precision>(r#""ns""#).unwrap(),
            Precision::Nanoseconds
        );
        assert_eq!(
            from_str::<Precision>(r#""u""#).unwrap(),
            Precision::Microseconds
        );
        assert_eq!(
            from_str::<Precision>(r#""ms""#).unwrap(),
            Precision::Milliseconds
        );
        assert_eq!(from_str::<Precision>(r#""s""#).unwrap(), Precision::Seconds);
        assert_eq!(from_str::<Precision>(r#""m""#).unwrap(), Precision::Minutes);
        assert_eq!(from_str::<Precision>(r#""h""#).unwrap(), Precision::Hours);
        assert_eq!(from_str::<Precision>(r#""d""#).unwrap(), Precision::Days);
        assert_eq!(from_str::<Precision>(r#""w""#).unwrap(), Precision::Weeks);

        // Test the alias for Microseconds
        assert_eq!(
            from_str::<Precision>(r#""µ""#).unwrap(),
            Precision::Microseconds
        );
    }

    #[test]
    fn test_precision_serialize() {
        // Test that each enum variant serializes to the expected string
        assert_eq!(to_string(&Precision::Nanoseconds).unwrap(), r#""ns""#);
        assert_eq!(to_string(&Precision::Microseconds).unwrap(), r#""u""#); // Note: serializes to "u", not "µ"
        assert_eq!(to_string(&Precision::Milliseconds).unwrap(), r#""ms""#);
        assert_eq!(to_string(&Precision::Seconds).unwrap(), r#""s""#);
        assert_eq!(to_string(&Precision::Minutes).unwrap(), r#""m""#);
        assert_eq!(to_string(&Precision::Hours).unwrap(), r#""h""#);
        assert_eq!(to_string(&Precision::Days).unwrap(), r#""d""#);
        assert_eq!(to_string(&Precision::Weeks).unwrap(), r#""w""#);
    }

    #[test]
    fn test_precision_error_cases() {
        // Test invalid inputs
        let invalid_result = from_str::<Precision>(r#""invalid""#);
        assert!(invalid_result.is_err());

        // Test case sensitivity (serde is case-sensitive by default)
        let uppercase_result = from_str::<Precision>(r#""NS""#);
        assert!(uppercase_result.is_err());
    }

    #[test]
    fn test_precision_roundtrip() {
        // Test serialization and deserialization roundtrip for all variants
        let variants = vec![
            Precision::Nanoseconds,
            Precision::Microseconds,
            Precision::Milliseconds,
            Precision::Seconds,
            Precision::Minutes,
            Precision::Hours,
            Precision::Days,
            Precision::Weeks,
        ];

        for variant in variants {
            let serialized = to_string(&variant).unwrap();
            let deserialized = from_str::<Precision>(&serialized).unwrap();
            assert_eq!(variant, deserialized);
        }
    }
}
