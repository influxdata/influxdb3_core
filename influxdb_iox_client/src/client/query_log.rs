use ::generated_types::google::{FieldViolation, OptionalField};
use client_util::connection::GrpcConnection;
use futures_util::{stream::BoxStream, StreamExt, TryStreamExt};

use self::generated_types::{query_log_service_client::QueryLogServiceClient, *};
use crate::connection::Connection;
use crate::error::Error;

/// Re-export generated_types
pub mod generated_types {
    pub use generated_types::influxdata::iox::querier::v1::*;
}

/// A basic client for working with the query log.
#[derive(Debug, Clone)]
pub struct Client {
    inner: QueryLogServiceClient<GrpcConnection>,
}

impl Client {
    /// Creates a new client with the provided connection
    pub fn new(connection: Connection) -> Self {
        Self {
            inner: QueryLogServiceClient::new(connection.into_grpc_connection()),
        }
    }

    /// Get log.
    pub async fn get_log(&mut self) -> Result<Log, Error> {
        let mut stream = self.inner.get_log(GetLogRequest {}).await?.into_inner();

        // first message should be metadata
        let Some(first_msg) = stream.try_next().await? else {
            return Err(Error::InvalidResponse(FieldViolation::required(
                "first message",
            )));
        };
        let get_log_response::Data::Metadata(metadata) = first_msg.data.unwrap_field("data")?
        else {
            return Err(Error::InvalidResponse(FieldViolation::required(
                "metadata in first message",
            )));
        };

        Ok(Log {
            metadata,
            entries: stream
                .map_err(Error::from)
                .and_then(|msg| async move {
                    let get_log_response::Data::Entry(entry) = msg.data.unwrap_field("data")?
                    else {
                        return Err(Error::InvalidResponse(FieldViolation::required(
                            "entry in tail messages",
                        )));
                    };
                    Ok(entry)
                })
                .boxed(),
        })
    }
}

/// [`get_log`](Client::get_log) response.
pub struct Log {
    /// Metadata.
    pub metadata: Metadata,

    /// Stream of entries.
    pub entries: BoxStream<'static, Result<LogEntry, Error>>,
}

impl std::fmt::Debug for Log {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Log")
            .field("metadata", &self.metadata)
            .field("entries", &"<STREAM>")
            .finish()
    }
}
