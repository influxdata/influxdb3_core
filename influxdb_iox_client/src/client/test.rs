use client_util::connection::GrpcConnection;
/// Re-export generated_types
use generated_types::{i_ox_testing_client::IOxTestingClient, TestErrorRequest};

use crate::connection::Connection;
use crate::error::Error;

/// A client for testing purposes
///
/// ```no_run
/// #[tokio::main]
/// # async fn main() {
/// use influxdb_iox_client::{
///     test::Client,
///     connection::Builder,
/// };
///
/// let mut connection = Builder::default()
///     .build("http://127.0.0.1:8082")
///     .await
///     .unwrap();
///
/// let mut client = Client::new(connection);
///
/// // trigger an error
/// client
///     .error()
///     .await
///     .expect("failed to trigger an error");
/// # }
/// ```
#[derive(Debug, Clone)]
pub struct Client {
    inner: IOxTestingClient<GrpcConnection>,
}

impl Client {
    /// Creates a new client with the provided connection
    pub fn new(connection: Connection) -> Self {
        Self {
            inner: IOxTestingClient::new(connection.into_grpc_connection()),
        }
    }

    /// Trigger an error.
    pub async fn error(&mut self) -> Result<(), Error> {
        let request = TestErrorRequest {};
        self.inner.test_error(request).await?;
        Ok(())
    }
}
