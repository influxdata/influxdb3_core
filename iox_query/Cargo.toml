[package]
name = "iox_query"
description = "IOx Query Interface and Executor"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[lints]
workspace = true

# This crate is designed to be independent of the rest of the IOx
# server and specific storage systems such as Mutable Buffer and Read Buffer.
#
# The rationale for this is to:
#
# 1. Keep change/compile/link time down during development when working on just this crate
# 2. Allow for query logic testing without bringing in all the storage systems.

[dependencies] # In alphabetical order
arrow = { workspace = true }
arrow_util = { path = "../arrow_util" }
async-trait = "0.1"
bytes = "1.6"
chrono = { version = "0.4", default-features = false }
data_types = { path = "../data_types" }
datafusion = { workspace = true }
datafusion_util = { path = "../datafusion_util" }
executor = { path = "../executor"}
futures = "0.3"
hashbrown = { workspace = true }
indexmap = { version = "2.2", features = ["std"] }
itertools = "0.12.1"
iox_query_params = { path = "../iox_query_params" }
iox_time = { path = "../iox_time" }
metric = { path = "../metric" }
object_store = { workspace = true }
observability_deps = { path = "../observability_deps" }
once_cell = "1"
parking_lot = "0.12"
parquet_file = { path = "../parquet_file" }
query_functions = { path = "../query_functions"}
schema = { path = "../schema" }
snafu = "0.8"
tokio = { version = "1.37", features = ["macros", "parking_lot"] }
tokio-stream = "0.1"
trace = { path = "../trace" }
tracker = { path = "../tracker" }
predicate = { path = "../predicate" }
uuid = { version = "1", features = ["v4"] }
workspace-hack = { version = "0.1", path = "../workspace-hack" }

[dev-dependencies] # In alphabetical order
test_helpers = { path = "../test_helpers" }
assert_matches = "1"
insta = { version = "1", features = ["yaml"] }
serde = { version = "1.0", features = ["derive"] }
