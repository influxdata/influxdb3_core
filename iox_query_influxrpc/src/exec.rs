pub(crate) mod fields_pivot_exec;
mod fields_pivot_stream;
mod series_pivot_exec;
mod series_pivot_stream;

pub(crate) use fields_pivot_exec::FieldsPivotExec;
pub(crate) use series_pivot_exec::SeriesPivotExec;
