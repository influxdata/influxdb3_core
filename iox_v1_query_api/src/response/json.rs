//! JSON encoding of InfluxQL query results.
use crate::types::Precision;

use super::{BufferedResponseStream, ChunkedResponseStream, Response, ResponseSerializer};
use bytes::buf::BufMut;
use bytes::{Bytes, BytesMut};
use futures::Stream;
use serde::Serialize;
use serde_json::ser::Serializer;
use std::{
    pin::Pin,
    task::{Context, Poll},
};
use tracing::warn;

/// A generic JSON-encoded [Response] stream.
pub(crate) struct JsonStream<S, F> {
    stream: S,
    formatter_fn: F,
    epoch: Option<Precision>,
}

impl<S, F> JsonStream<S, F> {
    pub(crate) fn new(stream: S, formatter_fn: F, epoch: Option<Precision>) -> Self {
        Self {
            stream,
            formatter_fn,
            epoch,
        }
    }

    fn poll_next_inner(&mut self, cx: &mut Context<'_>) -> Poll<Option<Response>>
    where
        S: Stream<Item = Response> + Unpin,
    {
        Pin::new(&mut self.stream).poll_next(cx)
    }
}

impl<S, F, Fmt> Stream for JsonStream<S, F>
where
    S: Stream<Item = Response> + Unpin,
    F: Fn() -> Fmt + Unpin,
    Fmt: serde_json::ser::Formatter,
{
    type Item = Bytes;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();
        this.poll_next_inner(cx).map(|opt| {
            opt.and_then(|resp| {
                let mut w = BytesMut::new().writer();
                let formatter = (this.formatter_fn)();
                let mut serializer = Serializer::with_formatter(&mut w, formatter);
                let resp = ResponseSerializer::new(&resp, this.epoch, false);
                if let Err(e) = resp.serialize(&mut serializer) {
                    warn!(error = %e, "failed to serialize response");
                    let mut w = BytesMut::new().writer();
                    let formatter = (this.formatter_fn)();
                    let mut serializer = Serializer::with_formatter(&mut w, formatter);
                    let error_field = serde_json::json!({
                        "error": format!("{}", e),
                    });
                    if let Err(e) = error_field.serialize(&mut serializer) {
                        warn!(error = %e, "failed to serialize error field");
                        return None;
                    }
                    Some(w.into_inner().freeze())
                } else {
                    Some(w.into_inner().freeze())
                }
            })
        })
    }
}

pub(crate) type ChunkedJsonStream<F> = JsonStream<ChunkedResponseStream, F>;
pub(crate) type BufferedJsonStream<F> = JsonStream<BufferedResponseStream, F>;

#[cfg(test)]
mod tests {
    use super::super::tests::{Column, make_statement};
    use super::*;

    use arrow::array::{
        ArrayRef, DictionaryArray, Float64Array, StringArray, TimestampNanosecondArray,
    };
    use arrow::datatypes::Int32Type;
    use futures::StreamExt;
    use iox_query::QueryDatabase;
    use iox_query::exec::IOxSessionContext;
    use iox_query::query_log::QueryLog;
    use iox_query::test::TestDatabaseStore;
    use iox_time::SystemProvider;
    use serde_json::ser::{CompactFormatter, PrettyFormatter};
    use std::sync::Arc;

    #[tokio::test]
    async fn empty_stream() {
        let stream = ChunkedResponseStream::new(vec![], 2);
        let mut chunked_json_stream = ChunkedJsonStream::new(stream, || CompactFormatter, None);
        assert!(chunked_json_stream.next().await.is_none());

        let stream = BufferedResponseStream::new(vec![]);
        let mut buffered_json_stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        assert!(buffered_json_stream.next().await.is_none());
    }

    #[tokio::test]
    async fn single_chunk() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![1.0, 2.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn single_chunk_pretty() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![1.0, 2.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedJsonStream::new(stream, PrettyFormatter::new, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, PrettyFormatter::new, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn single_chunk_exponential_value() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![
                73070599793680680000000000000000000000000000.0,
                73070599793680670000000000000000000000000000.0,
            ])),
        ];

        // Only test for buffered since the purpose is to test the formatting of the exponential value
        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn single_chunk_infinite_value() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![f64::INFINITY, f64::NEG_INFINITY])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedJsonStream::new(stream, PrettyFormatter::new, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, PrettyFormatter::new, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_chunks() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_chunks_many_measurements() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.5])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_statements() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns1 = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data1: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
        ];
        let columns2 = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: true,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data2: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1b", "t1b",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
        ];

        let statement1 = make_statement(&db, &ctx, &log, columns1.clone(), data1.clone());
        let statement2 = make_statement(&db, &ctx, &log, columns2.clone(), data2.clone());
        let stream = ChunkedResponseStream::new(vec![statement1, statement2], 2);
        let stream = ChunkedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement1 = make_statement(&db, &ctx, &log, columns1.clone(), data1.clone());
        let statement2 = make_statement(&db, &ctx, &log, columns2.clone(), data2.clone());
        let stream = BufferedResponseStream::new(vec![statement1, statement2]);
        let stream = BufferedJsonStream::new(stream, || CompactFormatter, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn test_epoch() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![1.0, 2.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream =
            ChunkedJsonStream::new(stream, || CompactFormatter, Some(Precision::Nanoseconds));
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream =
            BufferedJsonStream::new(stream, || CompactFormatter, Some(Precision::Nanoseconds));
        insta::assert_snapshot!(collect_output(stream).await);
    }

    async fn collect_output<S: Stream<Item = Bytes> + Send>(stream: S) -> String {
        String::from_utf8(
            stream
                .map(Vec::<u8>::from)
                .map(|mut v| {
                    v.push(b'\n');
                    v
                })
                .concat()
                .await,
        )
        .unwrap()
    }
}
