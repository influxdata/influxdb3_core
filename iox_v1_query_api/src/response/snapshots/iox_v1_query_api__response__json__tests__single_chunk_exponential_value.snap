---
source: iox_v1_query_api/src/response/json.rs
expression: collect_output(stream).await
---
{"results":[{"statement_id":0,"series":[{"name":"m1","columns":["time","t1","f1"],"values":[["1970-01-01T00:00:01Z","t1a",7.307059979368068e43],["1970-01-01T00:00:02Z","t1a",7.307059979368067e43]]}]}]}
